include/hip/hcc_detail/device_functions.h

```diff
@@ -142,72 +516,466 @@ CUDA implements half as unsigned short whereas, HIP doesn't.
-__device__ double __hiloint2double(int hi, int lo);
-__device__ double __int2double_rn(int x);
+// Clock functions
+__device__
+inline
+long long int __clock64() { return (long long int) __builtin_amdgcn_s_memrealtime(); }
 
-__device__ float __int2float_rd(int x);
-__device__ float __int2float_rn(int x);
-__device__ float __int2float_ru(int x);
-__device__ float __int2float_rz(int x);
+__device__
+inline
+long long int  __clock() { return (long long int) __builtin_amdgcn_s_memrealtime(); }
 
-__device__ float __int_as_float(int x);
+// hip.amdgcn.bc - named sync
+__device__
+inline
+void __named_sync(int a, int b) { __builtin_amdgcn_s_barrier(); }
 
-__device__ double __ll2double_rd(long long int x);
-__device__ double __ll2double_rn(long long int x);
-__device__ double __ll2double_ru(long long int x);
-__device__ double __ll2double_rz(long long int x);
+#endif // __HIP_DEVICE_COMPILE__
 
-__device__ float __ll2float_rd(long long int x);
-__device__ float __ll2float_rn(long long int x);
-__device__ float __ll2float_ru(long long int x);
-__device__ float __ll2float_rz(long long int x);
+// warp vote function __all __any __ballot
+extern "C" __device__ inline uint64_t __activelanemask_v4_b64_b1(unsigned int input) {
+    uint64_t output;
+    // define i64 @__activelanemask_v4_b64_b1(i32 %input) #5 {
+    //   %a = tail call i64 asm "v_cmp_ne_i32_e64 $0, 0, $1", "=s,v"(i32 %input) #9
+    //   ret i64 %a
+    // }
+    __asm("v_cmp_ne_i32_e64 %0, 0, %1" : "=s"(output) : "v"(input));
+    return output;
+}
 
-__device__ double __longlong_as_double(long long int x);
+__device__
+inline
+unsigned int __activelanecount_u32_b1(unsigned int input) {
+    return  __popcll(__activelanemask_v4_b64_b1(input));
+}
 
-__device__ double __uint2double_rn(int x);
+__device__
+inline
+int __all(int predicate) {
+    return __popcll(__activelanemask_v4_b64_b1(predicate)) == __activelanecount_u32_b1(1);
+}
+
+__device__
+inline
+int __any(int predicate) {
+#ifdef NVCC_COMPAT
+    if (__popcll(__activelanemask_v4_b64_b1(predicate)) != 0)
+        return 1;
+    else
+        return 0;
+#else
+    return __popcll(__activelanemask_v4_b64_b1(predicate));
+#endif
+}
+
+__device__
+inline
+unsigned long long int __ballot(int predicate) {
+    return __activelanemask_v4_b64_b1(predicate);
+}
+
+__device__
+inline
+unsigned long long int __ballot64(int predicate) {
+    return __activelanemask_v4_b64_b1(predicate);
+}
+
+// hip.amdgcn.bc - lanemask
+__device__
+inline
+int64_t  __lanemask_gt()
+{
+    int32_t activelane = __ockl_activelane_u32();
+    int64_t ballot = __ballot64(1);
+    if (activelane != 63) {
+        int64_t tmp = (~0UL) << (activelane + 1);
+        return tmp & ballot;
+    }
+    return 0;
+}
+
+__device__
+inline
+int64_t __lanemask_lt()
+{
+    int32_t activelane = __ockl_activelane_u32();
+    int64_t ballot = __ballot64(1);
+    if (activelane == 0)
+        return 0;
+    return ballot;
+}
+
+#ifdef __HIP_DEVICE_COMPILE__
```

Comment:
we still need the device functions to be declared for host compilation, therefore the code should be like this
`#ifdef __HIP_DEVICE_COMPILE__
__device__
inline
void* __get_dynamicgroupbaseptr()
{
    // Get group segment base pointer.
    return (char*)__local_to_generic(__to_local(__llvm_amdgcn_groupstaticsize()));
}
#else
__device__
void* __get_dynamicgroupbaseptr();
#endif // __HIP_DEVICE_COMPILE__
`

Reply:
Hi Sam, I've cherry-picked your fixes into this PR related to your review. Thanks!